components/fxa-client/src/telemetry.rs

```diff
@@ -2,12 +2,268 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
 
-pub use crate::http_client::ProfileResponse as Profile;
-use crate::{error::*, FirefoxAccount};
+use crate::{error::*, scopes, FirefoxAccount};
+use jwcrypto::{EncryptionAlgorithm, EncryptionParameters, Jwk};
+use rand_rccrypto::rand::seq::SliceRandom;
+use rc_crypto::rand;
 
 impl FirefoxAccount {
-    pub fn get_ecosystem_anon_id(&mut self) -> Result<Option<String>> {
+    /// Get the ecosystem anon id, generating it if necessary.
+    ///
+    /// **💾 This method alters the persisted account state.**
+    pub fn get_ecosystem_anon_id(&mut self) -> Result<String> {
+        self.get_ecosystem_anon_id_helper(true)
+    }
+
+    pub fn get_ecosystem_anon_id_helper(&mut self, generate_placeholder: bool) -> Result<String> {
         let profile = self.get_profile(false)?;
-        Ok(profile.ecosystem_anon_id)
+        // Default case: the ecosystem anon ID was generated during login.
+        if let Some(ecosystem_anon_id) = profile.ecosystem_anon_id {
+            return Ok(ecosystem_anon_id);
+        }
+        if !generate_placeholder {
+            return Err(ErrorKind::IllegalState("ecosystem_anon_id should be present").into());
+        }
+        // For older clients, we generate an ecosystem_user_id,
+        // persist it and then return ecosystem_anon_id.
+        let mut ecosystem_user_id = vec![0u8; 16];
```

Comment:
The specs for this have recently changed, we need to generate 32 bytes here rather than 16.